File: java/json-transform/src/main/java/co/nlighten/jsontransform/functions/TransformerFunctionFind.java
package co.nlighten.jsontransform.functions;

import co.nlighten.jsontransform.functions.common.*;

import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;

public class TransformerFunctionFind extends TransformerFunction {
    public TransformerFunctionFind() {
        super(FunctionDescription.of(
            Map.of(
            "by", ArgumentType.of(ArgType.Transformer).position(0)
            )));
    }
    @Override
    public CompletionStage<Object> apply(FunctionContext context) {
        var streamer = context.getJsonElementStreamer(null);
        if (streamer == null)
            return null;
        var by = context.getJsonElement("by", false);
        var index = new AtomicInteger(0);
        var adapter = context.getAdapter();
        return streamer.stream()
                .filter(item -> {
                    var condition = context.transformItem(by, item, index.getAndIncrement());
                    return adapter.isTruthy(condition);
                })
                .findFirst()
                .orElse(null);
    }
}

File: javascript/json-transform/src/functions/TransformerFunctionFind.ts
import TransformerFunction from "./common/TransformerFunction";
import { ArgType } from "./common/ArgType";
import FunctionContext from "./common/FunctionContext";
import { isTruthy } from "../JsonHelpers";

class TransformerFunctionFind extends TransformerFunction {
  constructor() {
    super({
      arguments: {
        by: { type: ArgType.Transformer, position: 0, defaultIsNull: true },
      },
    });
  }

  override async apply(context: FunctionContext): Promise<any> {
    const streamer = await context.getJsonElementStreamer(null);
    if (streamer == null) return null;

    const by = await context.getJsonElement("by", false);
    let index = 0;
    return await streamer.stream().find(async item => {
      const condition = await context.transformItem(by, item, index++);
      return isTruthy(condition);
    });
  }
}

export default TransformerFunctionFind;
